src/Libraries/RevitNodes/Elements/Subelement.cs

```diff
@@ -0,0 +1,193 @@
+        /// </summary>
+        public Category Category
+        {
+            get { return new Category(InternalSubelement.Category); }
+        }
+
+        #endregion
+
+        #region Parameters
+        /// <summary>
+        /// Obtain all parameter ids of the subelement.
+        /// </summary>
+        public IList<ElementId> GetAllParameters()
+        {
+            return InternalSubelement.GetAllParameters();
+        }
+
+        /// <summary>
+        /// Get the value of the subelement parameter.
+        /// </summary>
+        /// <param name="paramId">The parameter id to be retrieved, as ElementId or Int.</param>
+        /// <returns></returns>
+        public object GetParameterValue(object paramId)
+        {
+            ElementId idParam = paramId as ElementId;
+            if(idParam == null)
+            {
+                Int64 longValue = (Int64)paramId;
+                idParam = new ElementId(longValue);
+                if (idParam == null)
+                    return string.Empty;
+            }
+            return GetObjectValue( InternalSubelement.GetParameterValue(idParam) );
+        }
+
+        private object GetObjectValue(ParameterValue value)
+        {
+            IntegerParameterValue intValue = value as IntegerParameterValue;
+            if (intValue != null)
+                return intValue.Value;
+
+            DoubleParameterValue doubleValue = value as DoubleParameterValue;
+            if (doubleValue != null)
+                return doubleValue.Value;
+
+            StringParameterValue strValue = value as StringParameterValue;
+            if (strValue != null)
+                return strValue.Value;
+
+            ElementIdParameterValue elemidValue = value as ElementIdParameterValue;
+            if (elemidValue != null)
+                return elemidValue.Value;
+
+            return string.Empty;
+        }
+
+        /// <summary>
+        /// Set the value of the specified subelement parameter. 
+        /// </summary>
+        /// <param name="paramId">The parameter id to be set, as ElementId</param>
+        /// <param name="value">The new parameter value to be set.</param>
+        /// <returns></returns>
+        public Subelement SetParameterValue(ElementId paramId, ParameterValue value)
+        {
+            TransactionManager.Instance.EnsureInTransaction(DocumentManager.Instance.CurrentDBDocument);
+
+            InternalSubelement.SetParameterValue(paramId, value);
+
+            TransactionManager.Instance.TransactionTaskDone();
+            return this;
+        }
+
+        /// <summary>
+        /// Set the value of the specified subelement parameter. 
+        /// </summary>
+        /// <param name="paramId">The parameter id to be set.</param>
+        /// <param name="value">The new parameter value to be set.</param>
+        /// <returns></returns>
+        public Subelement SetDoubleParameterValue(Int64 paramId, double value)
```

Comment:
Why overload with specific types? Is there a reason for having it this way?

I ask because for elements we just have one node (function) exposed that takes an object and tries to infer the type from the parameter type, raising exception if incompatible...

Reply:
This is a workaround to the subelement parameters that are not directly accessible via the paramId in the API, unlike element. As a result, if we pass the object value alone, it would not be possible to correctly convert the value.

When we improve the Parameter API to work with subelement, this can be retired, assuming that is a mechanism to retire the node methods in D4R.